### DISCUSSION TITLE
HTML comments

### DISCUSSION THREAD

**Discussion by lostenderman** [lostenderman/markdown#143] [General]
### Problem description:

Currently, HTML Block and Inline Elements can be parsed in one of four ways:

1. `inlineHtmlTag`
2. `inlineHtmlComment`
3. `inputBlockHtmlElement`
4. `blockHtmlCommentBegin` 
    ... 
    `blockHtmlCommentEnd`

Compared to the previous implementation, however, the content of HTML comments is not further parsed.

This comment
```
<!-- *a* -->
```

would be previously parsed as
```
\markdownRendererBlockHtmlCommentBegin
\markdownRendererEmphasis{a}
\markdownRendererBlockHtmlCommentEnd
```

Similarly, this
```
a <!-- *a* -->
```
would be previously parsed as
```
a \markdownRendererInlineHtmlComment{\markdownRendererEmphasis{a}}
```

HTML block comments in CommonMark also include the additional content on the line containing the string `-->`.

This
```
<!-- *a* --> *b*
```

is  `<html_block>&lt;!-- *a* --&gt; *b*</html_block>` in CommonMark.

Shouldn't these be parsed as `inputBlockHtmlElement` with all their content verbatim?

1. valid comment
```
<!-- *a* -->
```

2. valid comment with additional content on the line containing the string `-->`
```
<!-- *a* --> *b*
```

Additionally, block HTML "comments" are not as restricted as inline HTML comments.

This is a valid HTML block:
```
<!-- 

--

--->
```
Though it does include `--` and ends with `-`.

Although these are not valid HTML comments, CommonMark still recognized them as HTML blocks, so parsing them as 
`inputBlockHtmlElement` would make more sense than an HTML comment.

1. invalid comment
```
<!-- *a* -- --->
```

2. invalid comment with additional content on the line containing the string `-->`
```
<!-- *a* -- ---> *b*
```

---

### Proposition:

Render the current `blockHtmlComment` as `inputBlockHtmlElement`.

#### Alternative 1:

Something like `inputBlockHtmlComment` could substitute the current HTML comment block set of macros. The content would also be parsed verbatim, excluding the opening `<!--` and closing `-->`. This macro would be generated only when the HTML block is a valid comment (possibly without additional content on the closing line).

#### Alternative 2:

Since no further parsing is done, all HTML is parsed as an `inlineHtmlTag` or `inputBlockHtmlElement`.

**Comment by Witiko** [comment]
Agreed. Let's get rid of the `blockHtmlComment` renderer and renderer prototype altogether and have all block HTML handled by `inputBlockHtmlElement`. If the grammar of CommonMark does not recognize proper block comments, there seems to be no sense in having a renderer for them and maintain an illusion.

**Comment by lostenderman** [comment]
Closed with commit 91088c146ffae3cfb818dc8f691a9178175d5489.
